### PR TITLE
logical and bitwise functions

### PR DESCRIPTION
**Pull request by mtar**
## Description
Fixes: #225 #226, improves #366 

Changes proposed:
- Implementation of the logical functions (and, not, or, not)
- Implementation of bitwise functions (invert/not, left/right shift)
- Removing binary_bit_op and other functions in operations.py introduced in #366
- infer tensor type from torch result instead of operands in binary_op.

## Type of change
[x] Bug fix (non-breaking change which fixes an issue)
[x] New feature (non-breaking change which adds functionality)
[ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
[ ] Documentation update

Are all split configurations tested and accounted for?
[x] yes [ ] no
Does this change require a documentation update outside of the changes proposed?
[ ] yes [x] no
Does this change modify the behaviour of other functions?
[ ] yes [x] no
Are there code practices which require justification?
[ ] yes [x] no

### CODE REVIEW

**Comment by coquelin77** [comment]
@mtar i am going to need you to merge the master with this one. you know this code better than i

File: heat/core/logical.py [hunk 1]
```diff
@@ -207,3 +208,98 @@ def local_any(t, *args, **kwargs):
...
+    boolean_tensor : tensor of type bool
+        Element-wise result of t1 AND t2.
+
+    Examples:
+    ---------
+    >>> ht.logical_and(ht.array([True, False]), ht.array([False, False]))
+    tensor([ False, False])
+    """
+    return operations.__binary_op(
+        torch.Tensor.__and__, types.bool(t1, device=t1.device), types.bool(t2, device=t2.device)
```
[the remark this answers is not in the document]
Comment (mtar) [reply heat/core/logical.py]: torch tensor's and/or is bitwise.

File: heat/core/operations.py [hunk 2]
```diff
@@ -262,17 +35,104 @@ def __binary_op(operation, t1, t2):
...
+            output_comm = t1.comm
+
+            # ToDo: Fine tuning in case of comm.size>t1.shape[t1.split]. Send torch tensors only to ranks, that will hold data.
+            if t1.split is not None:
+                if t1.shape[t1.split] == 1 and t1.comm.is_distributed():
+                    warnings.warn(
+                        "Broadcasting requires transferring data of first operator between MPI ranks!"
+                    )
+                    if t1.comm.rank > 0:
+                        t1._DNDarray__array = torch.zeros(
```
[the remark this answers is not in the document]
Comment (mtar) [reply heat/core/operations.py]: I'm not familiar with this part of the code. It seems to be a subject to change (issue #294).

[the remark this answers is not in the document]
Comment (coquelin77) [reply heat/core/operations.py]: noted. the diff said that it was from you but if its legacy then an exception can be made

**Pull request merged**